Workshop Materials/notebooks/Splink in Databricks - Expert Answer.py

```diff
@@ -0,0 +1,449 @@
+# MAGIC     - Check model makes intuitive sense using `linker.match_weights_chart()`
+# MAGIC 3. **Edges** - `linker.predict()` generates match scores between pairs of nodes (i.e. "edges"):
+# MAGIC     - How are the match scores distributed? Expect a bimodal distribution (definite matches + definite non-matches)
+# MAGIC     - If there are ambiguous scores (0.1 < `match_proability` < 0.9; `match_weight` close to zero) we can use the waterfall chart to investigate
+# MAGIC 6. **Clusters** - once we're happy with our model, we probably have a rough idea of what threshold match score to use to define a match (e.g. >0.9). Spot checking a sample of clusters with `cluster_studio_dashboard` we can check for a couple of things, the biggest risk being due to false positives (linking records that aren't genuine matches):
+# MAGIC     - Are there any "hairball" clusters that have become overly large because several genuine clusters have been merged?
+# MAGIC     - Ideal clusters will be fully-connected (every node is directly linked to every other) - are there clusters that appear to be two or more subclusters only connected by a single (suspicious) edge?
+
+# COMMAND ----------
+
+data.toPandas().head()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## 1. Nodes
+# MAGIC > **Explore the data with splink's `missingness` and `profile_columns` methods - this is an iterative process to identify transformations to make to the input data:**
+# MAGIC > - **Data cleaning/standardisation - making sure all values in a column are formatted the same way, handling nulls or anomalies**
+# MAGIC > - **Feature engineering - creating new derived columns that could be useful for blocking or making custom comparisons**
+# MAGIC 
+# MAGIC 
+# MAGIC Initial ideas for standardisation:
+# MAGIC - Clean up `country_of_residence` and `nationality`
+# MAGIC - Create `year_of_birth` for blocking/fuzzy comparisons
+# MAGIC - Split forenames into `forename1`, `forename2` etc.
+# MAGIC - Create `full_address` to include `premises` + `address_line_1` + `locality` + `region` + `postal_code`
+# MAGIC - Set all string columns to upper case - string comparator functions are typically case-sensitive
+# MAGIC - Create partial postcode columns (`area`, `district`, `sector`) - see image [here](https://user-images.githubusercontent.com/7570107/136946496-8769b06c-e4a6-488d-95f1-526946d96aa7.png)
+
+# COMMAND ----------
+
+import pyspark.sql.functions as f
+
+# STANDARDISED DATA
+
+# Names
+data_std = data\
+    .withColumn("forename1", f.split(f.upper(data['forenames']), ' ').getItem(1))\
+    .withColumn("forename2", f.split(f.upper(data['forenames']), ' ').getItem(2))\
+    .withColumn("forename3", f.split(f.upper(data['forenames']), ' ').getItem(3))\
+    .withColumn("forename4", f.split(f.upper(data['forenames']), ' ').getItem(4))\
+    .withColumn("names_arr", f.expr("array_sort(array(forename1, forename2, surname))"))
```

Comment:
This is a terrible idea - names array for different individuals will have elements in a different order, in fact for same individual in case of duplicates can have a very different order if there are typos.